python/mujoco_mpc/demos/direct/direct_optimizer.py

```diff
@@ -0,0 +1,1168 @@
+# Copyright 2023 DeepMind Technologies Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import mujoco
+import numpy as np
+from numpy import typing as npt
+from typing import Tuple
+
+
+# %%
+# differentiate mj_differentiatePos wrt qpos1, qpos2
+def diff_differentiatePos(
+    model: mujoco.MjModel, dt: float, qpos1: npt.ArrayLike, qpos2: npt.ArrayLike
+) -> Tuple[npt.ArrayLike, npt.ArrayLike]:
+  # initialize Jacobians
+  jac1 = np.zeros((model.nv, model.nv))
+  jac2 = np.zeros((model.nv, model.nv))
+
+  # loop over joints
+  for j in range(model.njnt):
+    # get address in qpos, qvel
+    padr = model.jnt_qposadr[j]
+    vadr = model.jnt_dofadr[j]
+
+    # joint type cases
+    match model.jnt_type[j]:
+      case mujoco.mjtJoint.mjJNT_FREE:
+        for i in range(3):
+          jac1[vadr + i, vadr + i] = -1.0 / dt
+          jac2[vadr + i, vadr + i] = 1.0 / dt
+        vadr += 3
+        padr += 3
+        blk1 = np.zeros((3, 3))
+        blk2 = np.zeros((3, 3))
+        mujoco.mjd_subQuat(
+            qpos2[padr : (padr + 4)], qpos1[padr : (padr + 4)], blk2, blk1
+        )
+        idx = slice(vadr, vadr + 3)
+        jac1[idx, idx] = blk1
+        jac2[idx, idx] = blk2
+      case mujoco.mjtJoint.mjJNT_BALL:
+        blk1 = np.zeros((3, 3))
+        blk2 = np.zeros((3, 3))
+        idxq = slice(padr, padr + 4)
+        mujoco.mjd_subQuat(qpos2[idxq], qpos1[idxq], blk2, blk1)
+        idxv = slice(vadr, vadr + 3)
+        jac1[idxv, idxv] = blk1
+        jac2[idxv, idxv] = blk2
+      case mujoco.mjtJoint.mjJNT_HINGE:
+        jac1[vadr, vadr] = -1.0 / dt
+        jac2[vadr, vadr] = 1.0 / dt
+      case mujoco.mjtJoint.mjJNT_SLIDE:
+        jac1[vadr, vadr] = -1.0 / dt
+        jac2[vadr, vadr] = 1.0 / dt
+
+  return jac1, jac2
+
+
+# %%
+# velocity and acceleration from configuration
+# v1 = (q1 - q0) / h
+# a1 = (v2 - v1) / h = (q2 - 2q1 + q0) / h^2
+def qpos_to_qvel_qacc(
+    model: mujoco.MjModel,
+    qpos: npt.ArrayLike,
+    horizon: int,
+) -> Tuple[npt.ArrayLike, npt.ArrayLike]:
+  qvel = np.zeros((model.nv, horizon))
+  qacc = np.zeros((model.nv, horizon))
```

Comment:
shouldn't qvel and qacc be of length horizon - 1 and horizon - 2?

Reply:
There are horizon - 1 velocities and horizon - 2 accelerations. However, the indexing is nicer this way with padding so indexing for the discrete-time index is consistent across trajectories.